### PR TITLE
Add test for `createInsertedHTMLStream`

### PR DESCRIPTION
**Pull request by Ethan-Arrowood**
Adds a test for `createInsertedHTMLStream` by validating it inserts and encodes the HTML string returned by the `getServerInsertedHTML` argument.

Closes NEXT-2750

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 753585be1419f4a62d69ba5f2b938891741e4bcf

`TURBOPACK=1 pnpm test test/integration/create-next-app/index.test.ts` (turbopack)
- create next app > valid example
- create next app > should allow example with GitHub URL
- create next app > should allow example with GitHub URL and example-path
- create next app > should use --example-path over the file path in the GitHub URL

<details>
<summary>Expand output</summary>

  ● create next app › valid example

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      80 |   for (const file of files) {
      81 |     try {
    > 82 |       expect(existsSync(resolve(projectRoot, file))).toBe(true)
         |                                                      ^
      83 |     } catch (err) {
      84 |       require('console').error(
      85 |         `missing expected file ${file}`,

      at toBe (integration/create-next-app/lib/utils.ts:82:54)
      at integration/create-next-app/index.test.ts:137:30
      at useTempDir (lib/use-temp-dir.ts:24:5)
      at Object.<anonymous> (integration/create-next-app/index.test.ts:121:5)

  ● create next app › should allow example with GitHub URL

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      80 |   for (const file of files) {
      81 |     try {
    > 82 |       expect(existsSync(resolve(projectRoot, file))).toBe(true)
         |                                                      ^
      83 |     } catch (err) {
      84 |       require('console').error(
      85 |         `missing expected file ${file}`,

      at toBe (integration/create-next-app/lib/utils.ts:82:54)
      at integration/create-next-app/index.test.ts:194:30
      at useTempDir (lib/use-temp-dir.ts:24:5)
      at Object.<anonymous> (integration/create-next-app/index.test.ts:177:5)

  ● create next app › should allow example with GitHub URL and example-path

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      80 |   for (const file of files) {
      81 |     try {
    > 82 |       expect(existsSync(resolve(projectRoot, file))).toBe(true)
         |                                                      ^
      83 |     } catch (err) {
      84 |       require('console').error(
      85 |         `missing expected file ${file}`,

      at toBe (integration/create-next-app/lib/utils.ts:82:54)
      at integration/create-next-app/index.test.ts:258:30
      at useTempDir (lib/use-temp-dir.ts:24:5)
      at Object.<anonymous> (integration/create-next-app/index.test.ts:239:5)

  ● create next app › should use --example-path over the file path in the GitHub URL

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      80 |   for (const file of files) {
      81 |     try {
    > 82 |       expect(existsSync(resolve(projectRoot, file))).toBe(true)
         |                                                      ^
      83 |     } catch (err) {
      84 |       require('console').error(
      85 |         `missing expected file ${file}`,

      at toBe (integration/create-next-app/lib/utils.ts:82:54)
      at integration/create-next-app/index.test.ts:291:30
      at useTempDir (lib/use-temp-dir.ts:24:5)
      at Object.<anonymous> (integration/create-next-app/index.test.ts:272:5)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js test-createInsertedHTMLStream | Change |
| - | - | - | - |
| buildDuration | 13.6s | 13.7s | N/A |
| buildDurationCached | 8.3s | 6.2s | N/A |
| nodeModulesSize | 198 MB | 198 MB | ⚠️ +21.3 kB |
| nextStartRea..uration (ms) | 428ms | 436ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test-createInsertedHTMLStream | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 30.5 kB | 30.5 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | N/A |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 242 B | ✓ |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 45.6 kB | 45.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test-createInsertedHTMLStream | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test-createInsertedHTMLStream | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.21 kB | 4.21 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.57 kB | 6.57 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test-createInsertedHTMLStream | Change |
| - | - | - | - |
| _buildManifest.js gzip | 481 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test-createInsertedHTMLStream | Change |
| - | - | - | - |
| index.html gzip | 529 B | 528 B | N/A |
| link.html gzip | 541 B | 540 B | N/A |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test-createInsertedHTMLStream | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.1 kB | 95.1 kB | N/A |
| page.js gzip | 3.06 kB | 3.07 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test-createInsertedHTMLStream | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 627 B | 624 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test-createInsertedHTMLStream | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | ✓ |
| app-page-exp..prod.js gzip | 96.5 kB | 96.5 kB | ✓ |
| app-page-tur..prod.js gzip | 98.2 kB | 98.2 kB | ✓ |
| app-page-tur..prod.js gzip | 92.7 kB | 92.7 kB | ✓ |
| app-page.run...dev.js gzip | 149 kB | 149 kB | ✓ |
| app-page.run..prod.js gzip | 91.2 kB | 91.2 kB | ✓ |
| app-route-ex...dev.js gzip | 21.3 kB | 21.3 kB | ✓ |
| app-route-ex..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| app-route.ru...dev.js gzip | 20.9 kB | 20.9 kB | ✓ |
| app-route.ru..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.52 kB | 9.52 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.8 kB | 9.8 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.52 kB | 9.52 kB | ✓ |
| pages-turbo...prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| pages.runtim...dev.js gzip | 22.9 kB | 22.9 kB | ✓ |
| pages.runtim..prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| server.runti..prod.js gzip | 50.5 kB | 50.5 kB | ✓ |
| Overall change | 947 kB | 947 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js test-createInsertedHTMLStream | Change |
| - | - | - | - |
| 0.pack gzip | 1.56 MB | 1.56 MB | N/A |
| index.pack gzip | 105 kB | 106 kB | ⚠️ +444 B |
| Overall change | 105 kB | 106 kB | ⚠️ +444 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 753585be1419f4a62d69ba5f2b938891741e4bcf

File: packages/next/src/server/stream-utils/node-web-streams-helper.test.ts [hunk 1]
```diff
@@ -179,7 +180,26 @@ describe('node-web-stream-helpers', () => {
...
-  describe('createInsertedHTMLStream', () => {})
+  describe('createInsertedHTMLStream', () => {
+    it('should insert data into the start of the transform', async () => {
+      const encoder = new TextEncoder()
+      function getServerInsertedHTML() {
+        return Promise.resolve('<foo>bar</foo>')
+      }
+      const input = new ReadableStream({
+        start(controller) {
+          controller.enqueue(encoder.encode('<fuzz>buzz</fuzz>'))
```
Comment (gnoff) [inline packages/next/src/server/stream-utils/node-web-streams-helper.test.ts:191]: I'd enqueue more than one chunk. what we expect is to reproduce the server inserted html on each chunk. This currently makes sense because we always precede this transform with one that creates a buffered chunk of an entire set of HTML that React is trying to flush. If we got rid of the giant buffer of html then this implementation would no longer serve it's purpose.

This stream helper's implementation is so simple I'm not sure it warrants testing at this level since the test doesn't really encode the semantic. There are more complicated streams helpers where these types of low level tests are necessary to fully exercise all the edge cases being handled within. I think you can land this if you think it's a good idea but I don't think you should feel compelled to

**Review by gnoff — APPROVED** [review]

**Comment by Ethan-Arrowood** [comment]
Closing in favor of testing functions actually used directly by `render.tsx` and `app-render.tsx`.
